cpp/src/arrow/engine/substrait/extension_set.cc

```diff
@@ -261,155 +389,390 @@ struct ExtensionIdRegistryImpl : ExtensionIdRegistry {
   }
 
-  Status CanRegisterFunction(Id id,
-                             const std::string& arrow_function_name) const override {
-    if (function_id_to_index_.find(id) != function_id_to_index_.end()) {
-      return Status::Invalid("Function id was already registered");
+  Status AddSubstraitAggregateToArrow(
+      Id substrait_function_id, SubstraitAggregateToArrow conversion_func) override {
+    if (parent_) {
+      ARROW_RETURN_NOT_OK(
+          parent_->CanAddSubstraitAggregateToArrow(substrait_function_id));
     }
-    if (function_name_to_index_.find(arrow_function_name) !=
-        function_name_to_index_.end()) {
-      return Status::Invalid("Function name was already registered");
+    return AddSubstraitToArrowFunc<SubstraitAggregateToArrow>(
+        substrait_function_id, std::move(conversion_func), &substrait_to_arrow_agg_);
+  }
+
+  template <typename ConverterType>
+  Status AddArrowToSubstraitFunc(std::string arrow_function_name, ConverterType converter,
+                                 std::unordered_map<std::string, ConverterType>* dest) {
+    auto add_result = dest->emplace(std::move(arrow_function_name), std::move(converter));
+    if (!add_result.second) {
+      return Status::Invalid(
+          "Failed to register Arrow to Substrait function converter for Arrow function ",
+          arrow_function_name, " because a converter already existed");
     }
     return Status::OK();
   }
 
-  Status RegisterFunction(Id id, std::string arrow_function_name) override {
-    DCHECK_EQ(function_ids_.size(), function_name_ptrs_.size());
+  Status AddArrowToSubstraitCall(std::string arrow_function_name,
+                                 ArrowToSubstraitCall converter) override {
+    if (parent_) {
+      ARROW_RETURN_NOT_OK(parent_->CanAddArrowToSubstraitCall(arrow_function_name));
+    }
+    return AddArrowToSubstraitFunc(std::move(arrow_function_name), converter,
+                                   &arrow_to_substrait_);
+  }
 
-    Id copied_id{*uris_.emplace(id.uri.to_string()).first,
-                 *names_.emplace(id.name.to_string()).first};
+  Status AddArrowToSubstraitAggregate(std::string arrow_function_name,
+                                      ArrowToSubstraitAggregate converter) override {
+    if (parent_) {
+      ARROW_RETURN_NOT_OK(parent_->CanAddArrowToSubstraitAggregate(arrow_function_name));
+    }
+    return AddArrowToSubstraitFunc(std::move(arrow_function_name), converter,
+                                   &arrow_to_substrait_agg_);
+  }
 
-    const std::string& copied_function_name{
-        *function_names_.emplace(std::move(arrow_function_name)).first};
+  Status CanAddArrowToSubstraitCall(const std::string& function_name) const override {
+    if (arrow_to_substrait_.find(function_name) != arrow_to_substrait_.end()) {
+      return Status::Invalid(
+          "Cannot register function converter because a converter already exists");
```

Comment:
should we add converter for `function_name` (add function_name in the message) already exists?

Reply:
I updated this error message to:

```
      return Status::Invalid("Cannot register function converter for Substrait id ",
                             substrait_function_id.ToString(),
                             " because a converter already exists");
```

does that address your concern?

Reply:
Yes this works.